### ISSUE TITLE
.Net AmazonS3Client.MultipartCopyAsync not redirecting to appropriate bucket when USEast1 provided as default region

### ISSUE THREAD

**Issue by MartyIce** [aws/aws-sdk-net#2547]
### Describe the bug

We rely on AmazonS3Client for S3, and set the RegionEndpoint to RegionEndpoint.USEast1.  This will cause AWS to redirect any S3 requests for buckets in other regions automatically.  Using AWSSDK.S3 v 3.7.8.8, this works correctly for both:

* CopyObjectAsync
* MultipartCopyAsync (which we use for files > 5GB)

However, when upgrading to AWSSDK.S3 v 3.7.103, we discovered:

* still works for CopyObjectAsync
* MultipartCopyAsync fails for files in different regions, with error:

```
The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.
``` 

Is this behavior something being deprecated out of the SDK (hopefully not)?

### Expected Behavior

When invoking MultipartCopyAsync, using AmazonS3Client with RegionEndpoint = RegionEndpoint.USEast1, the call should succeed regardless of the actual region location of the S3 bucket.

### Current Behavior

When invoking MultipartCopyAsync, using AmazonS3Client with RegionEndpoint = RegionEndpoint.USEast1, the call is failing with the error:
```
The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.
``` 

### Reproduction Steps

```
using Amazon.Runtime;
using Amazon.S3.Model;
using Amazon.S3;
using Amazon;

var cancel = CancellationToken.None;

// necessary keys and bucket info
var accessKey = "REPLACE";
var secretAccessKey = "REPLACE";
var srcBucket = "REPLACE";
var srcKey = "REPLACE";
var destBucket = "REPLACE";
var destKey = "REPLACE";

// Build client
var credentials = new BasicAWSCredentials(accessKey, secretAccessKey);

var s3Config = new AmazonS3Config
{
    RegionEndpoint = RegionEndpoint.USEast1, // USEast1 redirects requests to appropriate region
    ForcePathStyle = false // https://[%BUCKETNAME%].s3.amazonaws.com/[%KEY%]
};
var s3Client = new AmazonS3Client(credentials, s3Config);

// initiate multipart upload request
var initiateRequest = new InitiateMultipartUploadRequest
{
    BucketName = destBucket,
    Key = destKey
};

var initiateResponse = await s3Client.InitiateMultipartUploadAsync(initiateRequest, cancel);
var uploadId = initiateResponse.UploadId;

// Get object size
var metadataRequest = new GetObjectMetadataRequest
{
    BucketName = srcBucket,
    Key = srcKey
};
var metadataResponse = await s3Client.GetObjectMetadataAsync(metadataRequest, cancel);
var objectSize = metadataResponse.ContentLength;

// Do the uploading
// The official AWS documentation recommends a partsize of 5MB.
var partSize = 5 * (long)Math.Pow(2, 20); // 5 MB
var numparts = (int)Math.Ceiling(objectSize / (double)partSize);

var copyResponses = new List<CopyPartResponse>();
for (var i = 0; i < numparts; ++i)
{
    var position = i * partSize;
    var lastPosition = Math.Min(objectSize - 1, position + partSize - 1);
    var copyRequest = new CopyPartRequest
    {
        DestinationBucket = destBucket,
        DestinationKey = destKey,
        SourceBucket = srcBucket,
        SourceKey = srcKey,
        UploadId = uploadId,
        FirstByte = position,
        LastByte = lastPosition,
        PartNumber = i + 1
    };
    copyResponses.Add(await s3Client.CopyPartAsync(copyRequest, cancel));
}

// Complete upload
var completeRequest = new CompleteMultipartUploadRequest
{
    BucketName = destBucket,
    Key = destKey,
    UploadId = uploadId
};
completeRequest.AddPartETags(copyResponses);
var result = await s3Client.CompleteMultipartUploadAsync(completeRequest, cancel).ConfigureAwait(false);

Console.WriteLine($"Result: {result.HttpStatusCode}");
```

With v3.7.8.8:
![image](https://user-images.githubusercontent.com/5083039/221232386-0fe09b11-0e84-4c0e-98a1-554e310cdbbb.png)

With v3.7.103:
![image](https://user-images.githubusercontent.com/5083039/221232534-280004a1-e87a-43d3-be55-ff6b55dd0ae6.png)
 

### Possible Solution

Re-enable the "default region redirect" for MultipartCopyAsync (sorry, not familiar with actual code 😄 )

### AWS .NET SDK and/or Package version used

AWSSDK.S3 3.7.103

### Targeted .NET Platform

.Net Standard 2.0, .Net 7.0

### Operating System and version

Windows 10

**Comment by ashishdhingra** [comment]
Example for multipart upload copy: https://docs.aws.amazon.com/AmazonS3/latest/userguide/CopyingObjectsMPUapi.html

**Comment by MartyIce** [comment]
To add clarification - this issue occurs when the file is in a bucket located in a different region than what is provided on the client.  ie:

Client - USEast1 (should redirect)
Bucket - USWest2

**Comment by ashishdhingra** [comment]
Reproducible using customer's scenario and code.
**Steps:**
- Created bucket in `us-west-2` region.
- Uploaded file about 266MB in size.
- Created a new .NET console application (.NET 7) and added `AWSSDK.S3 v 3.7.8.8` package.
- Configured credentials in the `default` profile.
- Used the below code (kindly note I'm copying object in the same bucket):
```C#
using Amazon.S3.Model;
using Amazon.S3;
using Amazon;

var cancel = CancellationToken.None;

var srcBucket = "<<bucket-name-us-west-2>>";
var srcKey = "dev.zip";
var destBucket = "<<bucket-name-us-west-2>>";
var destKey = "dev-copy-1.zip";

var s3Config = new AmazonS3Config
{
    RegionEndpoint = RegionEndpoint.USEast1, // USEast1 redirects requests to appropriate region
    ForcePathStyle = false
};
var s3Client = new AmazonS3Client(s3Config);

// initiate multipart upload request
var initiateRequest = new InitiateMultipartUploadRequest
{
    BucketName = destBucket,
    Key = destKey
};

var initiateResponse = await s3Client.InitiateMultipartUploadAsync(initiateRequest, cancel);
var uploadId = initiateResponse.UploadId;

// Get object size
var metadataRequest = new GetObjectMetadataRequest
{
    BucketName = srcBucket,
    Key = srcKey
};
var metadataResponse = await s3Client.GetObjectMetadataAsync(metadataRequest, cancel);
var objectSize = metadataResponse.ContentLength;

// Do the uploading
// The official AWS documentation recommends a partsize of 5MB.
var partSize = 5 * (long)Math.Pow(2, 20); // 5 MB
var numparts = (int)Math.Ceiling(objectSize / (double)partSize);

var copyResponses = new List<CopyPartResponse>();
for (var i = 0; i < numparts; ++i)
{
    var position = i * partSize;
    var lastPosition = Math.Min(objectSize - 1, position + partSize - 1);
    var copyRequest = new CopyPartRequest
    {
        DestinationBucket = destBucket,
        DestinationKey = destKey,
        SourceBucket = srcBucket,
        SourceKey = srcKey,
        UploadId = uploadId,
        FirstByte = position,
        LastByte = lastPosition,
        PartNumber = i + 1
    };
    copyResponses.Add(await s3Client.CopyPartAsync(copyRequest, cancel));
}

// Complete upload
var completeRequest = new CompleteMultipartUploadRequest
{
    BucketName = destBucket,
    Key = destKey,
    UploadId = uploadId
};
completeRequest.AddPartETags(copyResponses);
var result = await s3Client.CompleteMultipartUploadAsync(completeRequest, cancel).ConfigureAwait(false);

Console.WriteLine($"Result: {result.HttpStatusCode}");
```

**RESULT:** The file gets copied successfully.

- Upgraded to `AWSSDK.S3 v3.7.103.16`.
- Re-ran the same code.
**RESULT:** Exception `The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.` is thrown while executing `CopyPartAsync()`.

**Comment by peterrsongg** [comment]
@MartyIce This has been fixed in S3 (3.7.103.19). Feel free to re-open this issue if the problem persists
